File: Data Structures & Algorithms/maximum-product-subarray/submission-9.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        curr_maxx, curr_minn = 1, 1
        res = max(nums)

        for num in nums:
            if num == 0:
                curr_maxx = 1 
                curr_minn = 1
                continue

            temp = curr_maxx * num
            curr_maxx = max(temp, num * curr_minn, num)
            curr_minn = min(temp, num * curr_minn, num)
            res = max(res, curr_maxx)
            
        return res

File: Data Structures & Algorithms/sliding-window-maximum/submission-9.go
func maxSlidingWindow(nums []int, k int) []int {
    q := []int{}
    res := []int{}

    for r := range nums {
        for len(q) > 0 && nums[q[len(q)-1]] < nums[r] {
            q = q[:len(q)-1]
        }

        q = append(q, r)

        if q[0] <= r-k {
            q = q[1:]
        }

        if r >= k-1 {
            res = append(res, nums[q[0]])
        }
    }

    return res
}